source/common/upstream/health_checker_impl.cc

```diff
@@ -334,11 +334,18 @@ void HttpHealthCheckerImpl::HttpActiveHealthCheckSession::onGoAway(
 
 HttpHealthCheckerImpl::HttpActiveHealthCheckSession::HealthCheckResult
 HttpHealthCheckerImpl::HttpActiveHealthCheckSession::healthCheckResult() {
-  uint64_t response_code = Http::Utility::getResponseStatus(*response_headers_);
+  const uint64_t response_code = Http::Utility::getResponseStatus(*response_headers_);
   ENVOY_CONN_LOG(debug, "hc response={} health_flags={}", *client_, response_code,
                  HostUtility::healthFlagsToString(*host_));
 
   if (!parent_.http_status_checker_.inRange(response_code)) {
+    // If the HTTP response code would indicate immediate failure AND the immediate health check
+    // failure header is set, exclude the host from LB.
+    // TODO(mattklein123):We could consider doing this check for any HTTP response code, but this
+    // seems like the least surprising behavior and we could consider relaxing this in the future.
+    if (response_headers_->EnvoyImmediateHealthCheckFail() != nullptr) {
+      host_->healthFlagSet(Host::HealthFlag::EXCLUDED_VIA_IMMEDIATE_HC_FAIL);
+    }
```

Comment:
Currently this will also not force a rebuild if the host was already failed. I should probably update this comment, but I'm not sure it's worth handling that case? @snowp WDYT?